src/DataStructures/VectorImpl.hpp

```diff
@@ -530,3 +530,78 @@ std::ostream& operator<<(std::ostream& os,
   MAKE_MATH_ASSIGN_EXPRESSION_POINTERVECTOR(-=, VECTOR_TYPE) \
   MAKE_MATH_ASSIGN_EXPRESSION_POINTERVECTOR(*=, VECTOR_TYPE) \
   MAKE_MATH_ASSIGN_EXPRESSION_POINTERVECTOR(/=, VECTOR_TYPE)
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief Defines the MakeWithValueImpl apply specialization
+ *
+ * \details The MakeWithValueImpl<type,ftype>::apply(type, type::value_type)
+ * constructed here produces an object with the same size as the type argument,
+ * initialized with the value argument in every entry.
+ *
+ * \param VECTOR_TYPE The vector type (e.g. `DataVector`)
+ */
+#define MAKE_WITH_VALUE_IMPL_DEFINITION_FOR(VECTOR_TYPE)  \
+  namespace MakeWithValueImpls {                          \
+  template <>                                             \
+  struct MakeWithValueImpl<VECTOR_TYPE, VECTOR_TYPE> {    \
+    static SPECTRE_ALWAYS_INLINE VECTOR_TYPE              \
+    apply(const VECTOR_TYPE& input,                       \
+          const VECTOR_TYPE::value_type value) noexcept { \
+      return VECTOR_TYPE(input.size(), value);            \
+    }                                                     \
+  };                                                      \
+  }  // namespace MakeWithValueImpls
+
+// {@
+/*!
+ * \ingroup DataStructuresGroup
+ * \ingroup TypeTraitsGroup
+ * \brief Helper struct to determine the element type of a VectorImpl or
+ * container of VectorImpl
+ *
+ * \details Extracts the element type of a `VectorImpl`, a std::array of
+ * `VectorImpl`, or a reference or pointer to a `VectorImpl`. In any of these
+ * cases, the `type` member is defined as the `ElementType` of the `VectorImpl`
+ * in question. If, instead, `get_vector_element_type` is passed an arithmetic
+ * or complex arithemetic type, the `type` member is defined as the passed type.
+ *
+ * \snippet DataStructures/Test_VectorImpl.cpp get_vector_element_type_example
+ */
+template <typename T, bool = cpp17::is_arithmetic_v<T>,
+          bool = static_cast<bool>(tt::is_a_v<std::complex, T>)>
```

Comment:
oddly enough, yes. Otherwise, the compiler believes the type can depend on the template argument, which violates specialization restrictions. The cast avoids that compiler error.

Reply:
Perhaps add a summarizing comment?

Reply:
@nilsdeppe is `tt::is_a_v` implemented correctly?  Shouldn't the second template argument be `typename T` to match the general `tt::is_a` . (Same for `tt::is_a_t`)